src/checkpoint/container/kokkos_unordered_map_serialize.h

```diff
@@ -0,0 +1,159 @@
+
+template <
+  typename Serializer, typename Key, typename Value, typename Device,
+  typename Hasher, typename EqualTo
+>
+void serializeKokkosUnorderedMapElems(
+  Serializer& s,
+  Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>& map
+) {
+  using mapSizeType =
+    typename Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>::size_type;
+
+  for (mapSizeType i = 0; i < map.capacity(); i++) {
+    Key keyAtI = map.key_at(i);
+    if (map.exists(keyAtI)) {
+      Value val = map.value_at(map.find(keyAtI));
+
+      s | keyAtI;
+      s | val;
+    }
+  }
+}
+
+template <
+  typename Serializer, typename Key, typename Value, typename Device,
+  typename Hasher, typename EqualTo
+>
+void deserializeInsertElems(
+  Serializer& s, Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>& map,
+  typename Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>::size_type
+    map_size,
+  typename Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>::size_type
+    map_capacity
+) {
+  using KeyAlloc = dispatch::Allocator<Key>;
+  using ValueAlloc = dispatch::Allocator<Value>;
+  using KeyReconstructor =
+    dispatch::Reconstructor<typename dispatch::CleanType<Key>::CleanT>;
+  using ValReconstructor =
+    dispatch::Reconstructor<typename dispatch::CleanType<Value>::CleanT>;
+
+  // resize unordered map
+  map.rehash(map_capacity);
+
+  KeyAlloc keyAllocated;
+  ValueAlloc valAllocated;
+  for (SerialSizeType i = 0; i < map_size; i++) {
+    auto* key = KeyReconstructor::construct(keyAllocated.buf);
+    auto* val = ValReconstructor::construct(valAllocated.buf);
+
+    s | *key;
+    s | *val;
+
+    map.insert(std::move(*key), std::move(*val));
+  }
+}
+
+template <
+  typename SerializerT, typename Key, typename Value, typename Device,
+  typename Hasher, typename EqualTo
+>
+typename std::enable_if_t<
+  not std::is_same<SerializerT, checkpoint::Footprinter>::value, void
+> serialize(
+  SerializerT& s,
+  Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>& map
+) {
+  using UnorderedMapType =
+    Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>;
+
+  typename UnorderedMapType::size_type size = serializeContainerSize(s, map);
+  typename UnorderedMapType::size_type capacity = serializeContainerCapacity(s, map);
+
+  if (s.isUnpacking()) {
+    deserializeInsertElems(s, map, size, capacity);
+  } else {
+    serializeKokkosUnorderedMapElems(s, map);
+  }
+}
+
+template <
+  typename SerializerT, typename Key, typename Value, typename Device,
+  typename Hasher, typename EqualTo
+>
+typename std::enable_if_t<
+  std::is_same<SerializerT, checkpoint::Footprinter>::value, void
+> serialize(
+  SerializerT& s,
+  Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>& map
+) {
+  using UnorderedMapType =
+    Kokkos::UnorderedMap<Key, Value, Device, Hasher, EqualTo>;
+
+  typename UnorderedMapType::size_type size = serializeContainerSize(s, map);
+  s.addBytes(size * (sizeof(Key) + sizeof(Value)));
```

Comment:
is it possible to take capacity into account to improve precision?
see `std::vector` for example:
https://github.com/DARMA-tasking/magistrate/blob/594584f5ab51c9d5610c6a608f5a94e3c81d7c7d/src/checkpoint/container/vector_serialize.h#L147-L155

Reply:
:+1: